design-proposals/klip-61-bootstrapped-table-subscriptions.md

```diff
@@ -0,0 +1,85 @@
+seamlessly combine the two. Adding bootstrapped `TABLE` subscriptions to ksqlDB achieves this purpose by allowing users to query their 
+data at rest up until now, while also subscribing to any later changes from data that is in motion. 
+
+Users can try to achieve this functionality by simply combining a `TABLE` pull query followed by a `TABLE` push query, but this 
+could result in gaps in the returned data since there are no guarantees that the push query will not miss any data after the 
+pull query is executed. Implementing this feature, we will keep track off the offsets under the hood for the user so that 
+we can guarantee there is no missing data. It will also be more clean and intuitive for users to have to only use one query
+to achieve this.
+
+## What is in scope
+
+- Queries should be supported on `TABLES` with `EMIT ALL` at the end.
+- Bootstrapped subscriptions should retrieve all past data and subscribe to any further changes.
+- Bootstrapped subscriptions should not terminate until the user quits them.
+- Bootstrapped subscriptions will have feature parity with push queries v2, such as filters and projections.
+
+## What is not in scope
+
+- Bootstrapped subscriptions will not be supported on `STREAMS`. Instead, we will throw an informative error that tells the
+user to use `SET 'auto.offset.reset'='earliest';` with a push query in order to get all past data in the `STREAM` while 
+simultaneously subscribing to any future updates. This is because unlike `TABLES` which have a current state that a pull query retrieves,
+there is no current state for a `STREAM`. 
+
+## Value/Return
+
+Adding this feature will ensure that users do not miss any data by trying to implement their own bootstrapped `TABLE` 
+subscription by combining a pull query and a push query. It also reduces the potential complexity by only needing a single
+query to get both the current state of a `TABLE` and any future changes to that `TABLE`.
+
+## Public APIS
+
+Users will be able to add an `EMIT ALL` clause at the end of their push query v2 on a table in order to indicate that they
+want to perform a bootstrapped `TABLE` subscription. The proposed syntax would then look this:
+
+```SQL
+SET 'ksql.query.push.v2.enabled' = 'true';
+SELECT select_expr [, ...]                 
+  FROM table
+  [ WHERE where_condition ]
+  [ AND and_condition ]
+  EMIT ALL;
+```
+
+## Design
+
+For a bootstrapped `TABLE` subscription, we will peek at the latest offset in the ksqlDB output topic using the kafka admin
+console. We will then save this offset and start the pull query, which can now take as long as needed since we have already
+saved the latest offset. Once the pull query finishes and returns the current state, we can now begin the subscription 
+and supply the saved offset as the starting subscription token to the push query in order to receive any future changes.
```

Comment:
Should we point out that there is still a race condition between getting the output topic end-offsets and running the pull query, that may result in "time travel" when the subscription starts (ie, the subscription might contain a few updates that happened _before_ the pull query(>

Reply:
Yes this is a valid point, we should bring up this limitation to make it clear, and also makes it clear what future work can be done